src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -136,26 +136,26 @@ object JoinIndexRule
    * @return replacement plan
    */
   private def getReplacementPlan(index: IndexLogEntry, plan: LogicalPlan): LogicalPlan = {
-    val bucketSpec = BucketSpec(
-      numBuckets = index.numBuckets,
-      bucketColumnNames = index.indexedColumns,
-      sortColumnNames = index.indexedColumns)
-
-    val location = new InMemoryFileIndex(spark, Seq(new Path(index.content.root)), Map(), None)
-    val relation = HadoopFsRelation(
-      location,
-      new StructType(),
-      index.schema,
-      Some(bucketSpec),
-      new ParquetFileFormat,
-      Map())(spark)
-
     // Here we can't replace the plan completely with the index. This will create problems.
     // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
     // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
     // lead to wrong results. So we only replace the base relation.
     plan transform {
       case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
+        val bucketSpec = BucketSpec(
+          numBuckets = index.numBuckets,
+          bucketColumnNames = index.indexedColumns,
+          sortColumnNames = index.indexedColumns)
+
+        val location = new InMemoryFileIndex(spark, Seq(new Path(index.content.root)), Map(), None)
+        val relation = HadoopFsRelation(
+          location,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
```

Comment:
Currently, we only have one extra column and your suggested code works fine. However, I think it is better if we rely on the source data schema to pick schema for an index usage. This way if we add more extra column(s) later; this code works as expected with no change required. 

There is another advantage in relying on the source data schema, at the query time, to prune columns from the index schema: Imagine the source data was partitioned on two columns: A and B with a path like:

- source/A=a1/B=b1

Therefore, these two columns are added to index schema as part of lineage. Now at the query time, if user passes the data path as `source/A=a1`, Column A will no longer be part of data schema. If a rule decides to use the index then by relying on source data schema we make sure column A is excluded from the index usage too, but if simply use `field.name != IndexConstants.DATA_FILE_NAME_COLUMN` then column A will be added to the logical relation node that is replaced by the index and the transformed query plan fails with analyze error as A was not originally present in the data schema, at the query time, and hence it was not resolved.

Therefore, I think it is better if we always rely on source data schema at the query time. 
@sezruby  Please let me know if above is clear and makes sense to you.

Reply:
Thanks for the detailed explanation, but it's a bit unclear to me. The current version also just uses `index.schema`, so it means the current version cannot handle the case you described? 
I could be wrong, but this schema of `HadoopFsRelation` seems for reading index data and `updatedOutput` will filter unused columns out..?

Reply:
As current version uses index.schema it has a similar issue, however I updated my comment as I realized this behavior is actually desired. The reason for that is if the source data at the query time uses a subset of the original source data then we should not use index as it could have extra rows that were part of original data and now excluded from the source data at query time. Imagine the index was created on source data with two partitions: `.../data/A=a1` and `.../data/A=a2`. An index created on `.../data` path has records from both partitions `a1` and `a2`. However, if data path at query time is set to `.../data/A=a1` then it means records from the `a2` partition should be excluded from the final results and if we use the index during query we may get those records back depending on the query and index config (i.e. potentially wrong query results). Moreover, our current signature computation/comparison does not mark such an index as a candidate for the query - and this is a correct behavior.
Regarding your point about HadoopFsRelation, if I understood what you mentioned correctly: Yes that schema is used for reading data; however any column in the schema is expected to be already resolved (i.e. has to have an `ExprId`) by the time we switch to an index. Once we switch to an index, we still use the same `ExprId` that Spark had assigned to the columns when analyzing the plan with source data. Hence, if the index schema has some column which was not present during former analyze phase, the query fails after transformation as such a column has no `ExprId`.

Reply:
Ok, I got the point. I might also need to investigate `ExprId` related things for hybrid scan. Thanks :)